### PR TITLE
8289071: Compute allocation sizes of stubs and nmethods outside of lock protection

### PR DESCRIPTION
**Pull request by yftsai**
The pattern of computing the allocation size is inconsistent among the derived classes of CodeBlob. These sizes are not based on the shared resources but computed inside lock-protected regions in several classes. This change moves these cases outside the protected region.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289071](https://bugs.openjdk.org/browse/JDK-8289071): Compute allocation sizes of stubs and nmethods outside of lock protection

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * @eastig (no known github.com user name / role)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9266/head:pull/9266` \
`$ git checkout pull/9266`

Update a local copy of the PR: \
`$ git checkout pull/9266` \
`$ git pull https://git.openjdk.org/jdk pull/9266/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9266`

View PR using the GUI difftool: \
`$ git pr show -t 9266`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9266.diff">https://git.openjdk.org/jdk/pull/9266.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/code/nmethod.cpp [hunk 1]
```diff
@@ -470,9 +470,9 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
   code_buffer->finalize_oop_references(method);
   // create nmethod
   nmethod* nm = NULL;
+  int native_nmethod_size = CodeBlob::allocation_size(code_buffer, sizeof(nmethod));
```
Comment (eastig) [inline src/hotspot/share/code/nmethod.cpp:473]: This change and another below are not calculation of stub sizes. You need either create another PR for them or update JDK-8289071 to include these cases.

**Comment by eastig** [comment]
Could you please add the description of your changes?
It would be very useful if the description briefly gives:
- What the problem is.
- What the solution is.
- How you tested them.

No need to copy all details from JDK-8289071, just enough to understand the context of the changes.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by eastig — APPROVED** [review]
lgtm

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by phohensee** [comment]
/sponsor

Presubmit test failure looks intentional.

**Pull request closed without merging**